scipy/optimize/_root_scalar.py

```diff
@@ -0,0 +1,236 @@
+"""
+Unified interfaces to root finding algorithms for real or complex
+scalar functions.
+
+Functions
+---------
+- root : find a root of a scalar function.
+"""
+from __future__ import division, print_function, absolute_import
+
+from scipy._lib.six import callable
+
+from . import zeros as optzeros
+
+__all__ = ['root_scalar']
+
+
+class MemoizeDer(object):
+    """ Decorator that caches the value and derivative(s) of function each
+    time it is called. """
+    def __init__(self, fun):
+        self.fun = fun
+        self.vals = None
+        self.x = None
+        self.nCalls = 0
```

Comment:
a couple of PEP8 violations slipped in, best to fix those: `n_calls` here, and `is_memoized` further down